internal/command/arguments/add_test.go

```diff
@@ -0,0 +1,159 @@
+package arguments
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/hashicorp/terraform/internal/addrs"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+)
+
+func TestParseAdd(t *testing.T) {
+	// need a pointer value for the -from-resource-addr tests
+	fromResource := mustResourceInstanceAddr("test_foo.bar")
+
+	tests := map[string]struct {
+		args      []string
+		want      *Add
+		wantError string
+	}{
+		"defaults": {
+			[]string{"test_foo.bar"},
+			&Add{
+				Addr:     mustResourceInstanceAddr("test_foo.bar"),
+				State:    &State{Lock: true},
+				ViewType: ViewHuman,
+			},
+			``,
+		},
+		"some flags": {
+			[]string{"-optional=true", "test_foo.bar"},
+			&Add{
+				Addr:     mustResourceInstanceAddr("test_foo.bar"),
+				State:    &State{Lock: true},
+				Optional: true,
+				ViewType: ViewHuman,
+			},
+			``,
+		},
+		"-from-state": {
+			[]string{"-from-state=test_foo.bar", "module.foo.test_foo.baz"},
```

Comment:
This scenario of using one resource instance to generate the config of another feels a little weird to me... I can imagine situations where it might make sense, like if it's a super-general resource that doesn't contain any unique identifiers that would conflict, but I guess my instinct here is to optimize for the common case of generating a configuration for the same address that we're reading from in the state, which would simplify this command line and remove an opportunity for a typo-related mistake:

```
terraform add -from-state module.foo.test_foo.baz
```

In this structure, the "from address" would be forced to be `module.foo.test_foo.baz`, matching the address of what we're generating.

I'm thinking here about the similar situation we had with `terraform import` historically, where you could import into something that wasn't in the configuration so if you made a typo then Terraform would plan to destroy the thing you just imported and create a new thing with the correct name. Allowing a different source address than the generated configuration seems like an opportunity for the same outcome from the opposite direction, which I'd rather avoid if we can. (Presumably the user can always change the name label in what Terraform generates here, if they really _do_ want a different name, right?)

What do you think? :thinking:

Reply:
I had identified this use-case all along (mostly for newer terraform adopters), and it's what I'd set out in the RFC, so I feel a little hesitant about changing it now. Someone could also conceivably want to use it to generate configuration for an entirely different workspace (not in the terraform workspace sense; an unrelated set of configuration files). 

Having said that, I'm speaking from my personal biases as a former terraform configuration author. I would have used `add` this way, so I assume that others will too. 🤷 

I'm willing to make this change, but I'm not entirely convinced - we can chat about this outside of GH.